api-put-object-fan-out.go

```diff
@@ -0,0 +1,145 @@
+/*
+ * MinIO Go Library for Amazon S3 Compatible Cloud Storage
+ * Copyright 2023 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package minio
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// PutObjectFanOutRequest this is the request structure sent
+// to the server to fan-out the stream to multiple objects.
+type PutObjectFanOutRequest struct {
+	Key                string            `json:"key"`
+	UserMetadata       map[string]string `json:"metadata,omitempty"`
+	UserTags           map[string]string `json:"tags,omitempty"`
+	ContentType        string            `json:"contentType,omitempty"`
+	ContentEncoding    string            `json:"contentEncoding,omitempty"`
+	ContentDisposition string            `json:"contentDisposition,omitempty"`
+	ContentLanguage    string            `json:"contentLanguage,omitempty"`
+	CacheControl       string            `json:"cacheControl,omitempty"`
+	Retention          RetentionMode     `json:"retention,omitempty"`
+	RetainUntilDate    *time.Time        `json:"retainUntil,omitempty"`
+}
+
+// PutObjectFanOutResponse this is the response structure sent
+// by the server upon success or failure for each object
+// fan-out keys. Additionally this response carries ETag,
+// VersionID and LastModified for each object fan-out.
+type PutObjectFanOutResponse struct {
+	Key          string     `json:"key"`
+	ETag         string     `json:"etag,omitempty"`
+	VersionID    string     `json:"versionId,omitempty"`
+	LastModified *time.Time `json:"lastModified,omitempty"`
+	Error        error      `json:"error,omitempty"`
+}
+
+// PutObjectFanOut - is a variant of PutObject instead of writing a single object from a single
+// stream multiple objects are written, defined via a list of PutObjectFanOutRequests. Each entry
+// in PutObjectFanOutRequest carries an object keyname and its relevant metadata if any. `Key` is
+// mandatory, rest of the other options in PutObjectFanOutRequest are optional.
+func (c *Client) PutObjectFanOut(ctx context.Context, bucket string, body io.Reader, fanOutReq []PutObjectFanOutRequest) ([]PutObjectFanOutResponse, error) {
```

Comment:
It should be possible to add Form elements. Maybe something like:

```suggestion

// PutObjectFanOutRequest contains parameters for PutObjectFanOut requests.
type PutObjectFanOutRequest struct {
	// Bucket is the destination bucket.
	Bucket  string
	
	// Targets contains all targets to fan out to. 
	Targets []FanOutTarget
	
	// Data should return the data to upload.
	Data io.Reader
	
	// UserData can contain additional fields that will be added to the form.
	UserData map[string]string
}

// PutObjectFanOut - is a variant of PutObject instead of writing a single object from a single
// stream multiple objects are written, defined via a list of PutObjectFanOutRequests. Each entry
// in PutObjectFanOutRequest carries an object keyname and its relevant metadata if any. `Key` is
// mandatory, rest of the other options in PutObjectFanOutRequest are optional.
func (c *Client) PutObjectFanOut(ctx context.Context, o PutObjectFanOutRequest) ([]PutObjectFanOutResponse, error) {
	policy := NewPostPolicy()
	policy.SetBucket(o.Bucket)
	policy.SetKey(strconv.FormatInt(time.Now().UnixNano(), 16))
	for k, v := range o.UserData {
		policy.SetUserData(strings.TrimPrefix(strings.ToLower(k), "amz-x-"), v)
	}
```

Reply:
The reason I kept it this way is to match input and output slices to be an exact match - putting all of them into a big struct makes it a bit wide IMO. 

We do not set anything on the form field itself as there is no value in adding that, these are directly added per object in the JSON value itself.

Reply:
It also means to match that we are sending JSON line by line instead of a single large JSON struct.

Reply:
Also would like the `bucket` to be an actual arg, and the body as well to be separate. To designate them as mandatory args not optional - just like all our other APIs

```
APIName(context, bucketname, objectname, io.Reader, opts)
```

In this case since object is configurable it is part of `opts...`

```
APIName(context, bucketname, io.Reader, opts...)
```